feat: Amazon Elastic Container Registry plugin

blakeromano opened the pull request:
### Issue # (if applicable)

Closes #1

### Reason for this change

<!--What is the bug or use case behind this change?-->
We want to visualize data surrounding ECR images including specifically Image Scan Results

### Description of changes

<!--What code changes did you make? Have you made any important design decisions?-->
This adds a new ECR plugin for the frontend and backend which results in a new tab in the entity page showing up (if configured with the correct annotation).

### Description of how you validated changes
I have run this locally with an AWS account with and without AWS Inspector enabled which both seem to render as expected.
<!--Have you added any unit tests and/or integration tests?-->

### Checklist

- [x] My code adheres to the [CONTRIBUTING GUIDE](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md)

---

_By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license_

niallthomson commented:
Thanks @blakeromano its awesome you can contribute this!

With RIV it will take me a little while to spin it up locally to properly check it out. Initially I see some changes we'll want to make for consistency with the other plugins:

1. We create our backend APIs so that it accepts the catalog entity then looks up the relevant annotation(s). This gives us an easy way to inherit access control from RBAC on the catalog entity this belongs to. [Example](https://github.com/awslabs/backstage-plugins-for-aws/blob/main/plugins/ecs/backend/src/service/router.ts#L46)
2. Our annotations right now are generally an ARN or set of tags. We use an ARN because it encapsulates account ID and region and that gives us consistent cross-region and cross-account behavior. I'm fine keeping the name as long as we can get it aligned with the next point.
3. Driving off (2) we'll want to implement the standard resource location pattern that can find AWS resources potentially across multiple accounts and regions. This is largely abstracted in to a common library. [Example](https://github.com/awslabs/backstage-plugins-for-aws/blob/c813b999fd5dcaf1c4246ab8a0c5a9d4084eaaef/plugins/codepipeline/backend/src/service/DefaultAwsCodePipelineService.ts#L218)

blakeromano commented:
@niallthomson Thanks for the initial feedback!

I updated it so it is now using the catalog entity and does the lookup that way now RBAC should be respected.

I also updated to use an ARN rather than a name. From my understanding ECR API essentially is based off the registryId (which is the accountID), the region and then the repository name so I'll look more into the resource location pattern from the common library but at least it is using the ARN for now so at least there is alignment on using ARN versus having a one off on the name usage for now.

blakeromano commented on plugins/ecr/backend/src/service/router.ts:
```diff
@@ -0,0 +1,182 @@
+    router.use((req, res, next) => {
+      const cacheKey = req.originalUrl;
+
+      console.log(`Cache key ${cacheKey}`);
```
This also is https://github.com/awslabs/backstage-plugins-for-aws/blob/main/plugins/cost-insights/backend/src/service/router.ts#L58 but I do agree I moved it to a debug log with `logger`

blakeromano commented on plugins/ecr/backend/src/service/EcrAwsService.ts:
```diff
@@ -0,0 +1,233 @@
+      ECR_ARN_ANNOTATION,
+    ]);
+
+    const images = await this.ecrClient.send(
```
I think I agree with you my main thoughts are:

1. How easy do you make it to blow up the UI or by the backend by reading a ton of data into either the server memory?
2. It kind of feels like pagination in the frontend may make the most sense but I think would require a decent amount of rethinking of frontend logic.


I think I'll implement 1 for now as I think to your point that makes the most sense to me and is the easiest to implement.

niallthomson replied:
In the CodePipeline plugin we just [hard-coded the limit to 100](https://github.com/awslabs/backstage-plugins-for-aws/blob/97e23b7d220bdec38a2fb59764d741ea3b5cd2d6/plugins/codepipeline/backend/src/service/DefaultAwsCodePipelineService.ts#L58) executions for now.

I'm happy with this also being hard-coded to the same number and we open a different issue to figure out a consistent way to handle configuring these.

blakeromano commented:
@fabbazon Thanks for comments I think I addressed everything you mentioned.

niallthomson commented on examples/entities.yaml:
```diff
@@ -12,11 +12,13 @@ apiVersion: backstage.io/v1alpha1
     aws.amazon.com/aws-codebuild-project-tags: component=example-website
     aws.amazon.com/amazon-ecs-service-tags: component=example-website
     aws.amazon.com/cost-insights-tags: component=example-website
+    aws.amazon.com/aws-ecr-repository-arn: 1234567890.dkr.ecr.us-east-1.amazonaws.com/example-website
```
Hmmmm, I want this to work with this repo and the example CDK without any modification as all the other examples do.

Think we have two options:
1. Add back in an annotation `aws-ecr-repository-name`, which doesn't have an equivalent in the other plugins
2. Support tags. The backend code for this is relatively simple because we've abstracted the hard work. The frontend gets annoying because you have to be able to handle potentially getting multiple repositories back.

niallthomson commented:
I just pushed a bunch of changes to basically make this consistent with the other plugins in terms of:

1. Annotations
2. UI design
3. General naming in the code

It now works with annotations for both ARN and tags the same as plugins like CodePipeline does.

The main view is now oriented around ECR repositories and listing the images. You can click the "Findings" column for an image to get the detail view of the findings as a table, then navigate back.

Some things that I removed and we can put back in either now or later:

1. Caching: The way it was implemented would get funky with RBAC implementations, since it wouldn't account for whether a user could access a given catalog entity. Raised issue #328 to consider this across all the plugins but suggest we skip for now.
2. Findings gauges: I want to get this added back in since its useful but will slightly tweak how the UI looks so its consistent with the ECR console.
